### PR TITLE
WW-5352 Refactor ParametersInterceptor

### PR DESCRIPTION
**Pull request by kusalk**
WW-5352
--
General clean up in preparation for #832

### CODE REVIEW

File: core/src/main/java/com/opensymphony/xwork2/security/AcceptedPatternsChecker.java [hunk 1]
```diff
@@ -32,37 +32,37 @@ public interface AcceptedPatternsChecker {
      * @param value to check
      * @return object containing result of matched pattern and pattern itself
      */
-    public IsAccepted isAccepted(String value);
```
Comment (kusalk) [inline core/src/main/java/com/opensymphony/xwork2/security/AcceptedPatternsChecker.java:35]: Redundant in interfaces

File: core/src/main/java/org/apache/struts2/dispatcher/Parameter.java [hunk 2]
```diff
@@ -58,7 +58,7 @@ public String getName() {
         @Override
         public String getValue() {
             String[] values = toStringArray();
-            return (values != null && values.length > 0) ? values[0] : null;
```
Comment (kusalk) [inline core/src/main/java/org/apache/struts2/dispatcher/Parameter.java:61]: `#toStringArray` never returns a null value

File: core/src/main/java/org/apache/struts2/interceptor/parameter/ParametersInterceptor.java [hunk 3]
```diff
@@ -132,138 +133,149 @@ static private int countOGNLCharacters(String s) {
     @Override
     public String doIntercept(ActionInvocation invocation) throws Exception {
```
Comment (kusalk) [inline core/src/main/java/org/apache/struts2/interceptor/parameter/ParametersInterceptor.java:134]: Used early returns to reduce nesting and improve readability

File: core/src/main/java/org/apache/struts2/interceptor/parameter/ParametersInterceptor.java [hunk 4]
```diff
@@ -132,138 +133,149 @@ static private int countOGNLCharacters(String s) {
...
      * @param newParams The parameter map to apply
-     *                  <p>
-     *                  In this class this is a no-op, since the parameters were fetched from the same location.
-     *                  In subclasses both retrieveParameters() and addParametersToContext() should be overridden.
-     *                  </p>
      */
     protected void addParametersToContext(ActionContext ac, Map<String, ?> newParams) {
     }
 
     protected void setParameters(final Object action, ValueStack stack, HttpParameters parameters) {
```
Comment (kusalk) [inline core/src/main/java/org/apache/struts2/interceptor/parameter/ParametersInterceptor.java:185]: Extracted a bunch of helper methods to make this method much more readable

File: core/src/main/java/org/apache/struts2/interceptor/parameter/ParametersInterceptor.java [hunk 5]
```diff
@@ -287,13 +302,11 @@ protected boolean isAcceptableParameter(String name, Object action) {
      * @return true if parameter is accepted
      */
     protected boolean isAcceptableParameterValue(Parameter param, Object action) {
-        ParameterValueAware parameterValueAware = (action instanceof ParameterValueAware) ? (ParameterValueAware) action : null;
-        boolean acceptableParamValue = (parameterValueAware == null || parameterValueAware.acceptableParameterValue(param.getValue()));
-        if (hasParamValuesToExclude() || hasParamValuesToAccept()) {
-            // Additional validations to process
-            acceptableParamValue &= acceptableValue(param.getName(), param.getValue());
```
Comment (kusalk) [inline core/src/main/java/org/apache/struts2/interceptor/parameter/ParametersInterceptor.java:294]: `&=` doesn't short-circuit

File: core/src/main/java/org/apache/struts2/interceptor/parameter/ParametersInterceptor.java [hunk 6]
```diff
@@ -338,18 +344,17 @@ protected boolean acceptableName(String name) {
             return false;
         }
         boolean accepted = isWithinLengthLimit(name) && !isExcluded(name) && isAccepted(name);
-        if (devMode && accepted) { // notify only when in devMode
```
Comment (kusalk) [inline core/src/main/java/org/apache/struts2/interceptor/parameter/ParametersInterceptor.java:341]: Cleaned out comments for behaviours which are fairly obvious from code

File: core/src/main/java/org/apache/struts2/interceptor/parameter/ParametersInterceptor.java [hunk 7]
```diff
@@ -132,138 +133,149 @@ static private int countOGNLCharacters(String s) {
...
-            ReflectionContextState.setDenyMethodExecution(context, true);
-            ReflectionContextState.setReportingConversionErrors(context, true);
-
-            //keep locale from original context
             newStack.getActionContext().withLocale(stack.getActionContext().getLocale()).withValueStack(stack);
         }
+        return newStack;
+    }
+
+    protected void setMemberAccessProperties(ValueStack stack) {
```
Comment (lukaszlenart) [inline core/src/main/java/org/apache/struts2/interceptor/parameter/ParametersInterceptor.java:216]: I would avoid using _setter/getter_ pattern if this isn't a setter _per se_ - this also reduces a risk of using such method in expression evaluation. What about using `applyMemberAccessProperties()` instead?

File: core/src/main/java/org/apache/struts2/interceptor/parameter/ParametersInterceptor.java [hunk 8]
```diff
@@ -132,138 +133,149 @@ static private int countOGNLCharacters(String s) {
...
+            return HttpParameters.create().withComparator(getOrderedComparator()).withParent(parameters).build();
+        } else {
+            return HttpParameters.create().withParent(parameters).build();
+        }
+    }
 
-        for (Map.Entry<String, Parameter> entry : acceptableParameters.entrySet()) {
-            String name = entry.getKey();
-            Parameter value = entry.getValue();
+    protected void setParametersOnStack(ValueStack stack, Map<String, Parameter> parameters, Object action) {
```
Comment (lukaszlenart) [inline core/src/main/java/org/apache/struts2/interceptor/parameter/ParametersInterceptor.java:254]: `applyParametersOnStack`?

File: core/src/main/java/org/apache/struts2/interceptor/parameter/ParametersInterceptor.java (hunk 7 above)
Comment (kusalk) [reply core/src/main/java/org/apache/struts2/interceptor/parameter/ParametersInterceptor.java:216]: I'll rename all of these :) They are `protected` so shouldn't be accessible from expressions but I agree with the premise

**Review by lukaszlenart — APPROVED** [review]

**Pull request merged**